File: projecteuler/four.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

__author__ = 'zhuwei'

"""
Problem 4
A palindromic number reads the same both ways.

The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.

"""


def is_palindrome2(num):
    # str_num = str(num)
    str_num = "%d" % num
    str_len = len(str_num)
    for i in xrange(0, str_len / 2):
        if str_num[i] != str_num[str_len - 1 - i]:
            return False
    return True


def is_palindrome(num):
    if num < 0:
        return False
    divisor = 1
    while num / divisor >= 10:
        divisor *= 10   # 获取num的有几位，比如999对应divisor为100
    while num != 0:
        left = num / divisor
        right = num % 10
        if left != right:
            return False
        num = (num % divisor) / 10  # 去除最高位和最低位,divisor对应缩小100倍
        divisor /= 100
    return True


def max_three_digit_product_palindrome():
    max_answer = 1
    for i in xrange(999, 99, -1):
        for j in xrange(999, 99, -1):
            if is_palindrome2(i * j):
                if i * j > max_answer:
                    max_answer = i * j
    return max_answer


if __name__ == '__main__':

    print max_three_digit_product_palindrome()

File: projecteuler/four.erl
#!/usr/bin/env escript
%%%-------------------------------------------------------------------
%%% @author zhuwei
%%% @copyright (C) 2015, <COMPANY>
%%% @doc
%%%
%%% @end
%%% Created : 30. 七月 2015 下午1:57
%%%-------------------------------------------------------------------
-module(four).
-author("zhuwei").

%% API
-export([]).

max_divisor(Number) ->
  max_divisor(Number, 1).

max_divisor(Number, Div) ->
  case round(Number div Div) >= 10 of
    'false' -> Div;
    'true' -> max_divisor(Number, Div * 10)
  end.


is_palindrome(Number) when Number < 0 ->
  'false';
is_palindrome(Number) ->
  MaxDiv = max_divisor(Number),
  is_palindrome(Number, MaxDiv).

is_palindrome(Number, _Div) when Number =:= 0 ->
  'true';
is_palindrome(Number, Div) ->
  Left = round(Number div Div),
  Right = Number rem 10,
  if Left =/= Right -> 'false';
    'true' -> is_palindrome(round((Number rem Div) div 10), round(Div div 100))
  end.


max_three_digit_product_palindrome(I, J) ->
  max_three_digit_product_palindrome(I, J, 1, J).

max_three_digit_product_palindrome(0, 0, MaxAnswer, OrgJ) ->
  MaxAnswer;
max_three_digit_product_palindrome(I, 0, MaxAnswer, OrgJ) ->
  max_three_digit_product_palindrome(I - 1, OrgJ, MaxAnswer, OrgJ);
max_three_digit_product_palindrome(I, J, MaxAnswer, OrgJ) ->
  case is_palindrome(I * J) of
    'true' ->
        if (I * J > MaxAnswer) -> max_three_digit_product_palindrome(I, J - 1, I * J, OrgJ);
          'true' -> max_three_digit_product_palindrome(I, J - 1, MaxAnswer, OrgJ)
        end;
    'false' ->
      max_three_digit_product_palindrome(I, J - 1, MaxAnswer, OrgJ)
  end.



main(_) ->
  io:format("999: ~p~n", [is_palindrome(999)]),
  io:format("191: ~p~n", [is_palindrome(191)]),
  io:format("299: ~p~n", [is_palindrome(299)]),
  io:format("906609: = ~p~n", [max_three_digit_product_palindrome(999, 999)]),
  io:format("22: ~p~n", [is_palindrome(22)]).



